qiskit/pulse/compiler/passes/set_sequence.py

```diff
@@ -0,0 +1,152 @@
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Sequencing pass for Qiskit PulseIR compilation."""
+
+from __future__ import annotations
+from functools import singledispatchmethod
+from rustworkx import PyDAG
+
+from qiskit.pulse.compiler.basepasses import TransformationPass
+from qiskit.pulse.ir import SequenceIR
+from qiskit.pulse.model import MixedFrame
+from qiskit.pulse.transforms import AlignmentKind, SequentialAlignment, ParallelAlignment
+from qiskit.pulse.exceptions import PulseCompilerError
+
+
+class SetSequence(TransformationPass):
+    """Sets the sequence of a :class:`.SequenceIR` object.
+
+    The pass traverses the :class:`.SequenceIR` and recursively sets the sequence, by adding edges to
+    the ``sequence`` property. Sequencing is done according to the alignment strategy.
+
+    For parallel alignment types, the pass depends on the results of the analysis pass
+    :class:`~qiskit.pulse.compiler.passes.MapMixedFrame`.
+    """
+
+    def __init__(self):
+        """Create new SetSequence pass"""
+        super().__init__(target=None)
+
+    def run(
+        self,
+        passmanager_ir: SequenceIR,
+    ) -> SequenceIR:
+        """Run sequencing pass.
+
+        Arguments:
+            passmanager_ir: The IR object to be sequenced.
+        """
+
+        self._sequence_instructions(passmanager_ir.alignment, passmanager_ir.sequence)
+        return passmanager_ir
+
+    @singledispatchmethod
+    def _sequence_instructions(self, alignment: AlignmentKind, sequence: PyDAG):
+        """Finalize the sequence by adding edges to the DAG
+
+        ``sequence`` is mutated to include all the edges
+        connecting the elements of the sequence.
+
+        Nested :class:`.SequenceIR` objects are sequenced recursively.
+        """
+        raise NotImplementedError
+
+    # pylint: disable=unused-argument
+    @_sequence_instructions.register(ParallelAlignment)
+    def _sequence_instructions_parallel(self, alignment: ParallelAlignment, sequence: PyDAG):
+        """Finalize the sequence by adding edges to the DAG, following parallel alignment.
+
+        ``sequence`` is mutated to include all the edges
+        connecting the elements of the sequence in parallel.
+
+        Nested :class:`.SequenceIR` objects are sequenced recursively.
+
+        Args:
+            alignment: The IR alignment.
+            sequence: The graph object to be sequenced.
+
+        Raises:
+            PulseCompilerError: if ``property_set`` does not include a mixed_frames_mapping dictionary.
+        """
+        if self.property_set["mixed_frames_mapping"] is None:
```

Comment:
Input type dependent check would be annoying for users. I would move to the `run` method and say the pass is dependent of `MapMixedFrame` in the `.. notes::` in the class doc (that's fair because sequential alignment context may contain internal parallel alignment context).